genie-agent/src/main/java/com/netflix/genie/agent/jobexecution/statemachine/stateactions/StateAction.java

```diff
@@ -0,0 +1,85 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.jobexecution.statemachine.stateactions;
+
+import com.netflix.genie.agent.jobexecution.statemachine.Events;
+import com.netflix.genie.agent.jobexecution.statemachine.States;
+import org.springframework.statemachine.action.Action;
+
+/**
+ * Interface types for individual state actions.
+ * @author mprimi
+ * @since 4.0.0
+ */
+public interface StateAction extends Action<States, Events> {
+
+    /**
+     * INITIALIZE state action interface.
+     */
+    interface Initialize extends StateAction {
```

Comment:
You already have `BaseAction` implementing `StateAction` but then you have these implementations again extending these classes. It seems circular and I wonder if theres some way out of it. `Default` method for an interface and then just have these extend `BaseAction`?

Reply:
The interfaces are for type safety and having an explicit bean type per action.
The implementations all need to extend `Base`.
But I agree this is a bit convoluted, will talk to you in person about it.